### ISSUE TITLE
ndfilters.convolve wrap mode not  working when the array is chunked

### ISSUE THREAD

**Issue by anlavandier** [dask/dask-image#242]
**What happened**: 
Used `ndfilters.convolve` using the `wrap` mode and the output had the wrong values on the edges
**What you expected to happen**:
The function to output the correct values on the edges. 
**Minimal Complete Verifiable Example**:

```python
import numpy as np
import dask.array as da
from dask_image.ndfilters import convolve


def test_correct(F,axis = 0):
    if axis !=0:
        F = da.swapaxes(F, 0, axis)
        result = test_correct(F,axis = 0)
        return result
    
    kernel = np.array([1,-1,0,1,-1]).reshape((5,) + (1,) * (F.ndim - 1))
    # The output using ndfilters.convolve with the wrap mode
    O1 = convolve(F,kernel, mode = 'wrap')
    # Output created by forcing the wrapping using dask.array.overlap.overlap then calculating the convolution. 
    # taking care of the boundary
    F_1 = da.overlap.overlap(F, depth = {axis : 2}, boundary = 'periodic')
    # calculate the convolution
    O2_int = convolve(F_1,kernel, mode = 'constant')
    # removing the useless ghosts zones. 
    O2 = da.overlap.trim_overlap(O2_int, depth = {axis : 2})

    return O1.compute(), O2.compute()

test_1D_no_chunks = da.from_array([1,2,3,4,5,6,7,8,9],chunks = (-1,))
test_1D_chunked = da.from_array([1,2,3,4,5,6,7,8,9],chunks = (5,))

print('1D no chunks :',*test_correct(test_1D_no_chunks), sep = '\n', end = '\n\n')
print('1D chunked :', *test_correct(test_1D_chunked), sep = '\n', end = '\n\n')

test_2D_no_chunks = da.from_array(np.arange(25).reshape(5,5),chunks =(-1,-1))
test_2D_chunked = da.from_array(np.arange(25).reshape(5,5),chunks =(-1,2))

print('2D no chunks axis 0  :',*test_correct(test_2D_no_chunks,axis = 0), sep = '\n\n', end = '\n\n\n')
print('2D no chunks axis 1  :',*test_correct(test_2D_no_chunks,axis = 1), sep = '\n\n', end = '\n\n\n')

print('2D chunked axis 0  :',*test_correct(test_2D_chunked,axis = 0), sep = '\n\n', end = '\n\n\n')
print('2D chunked axis 1  :',*test_correct(test_2D_chunked,axis = 1), sep = '\n\n', end = '\n\n\n')
```
```
1D no chunks :
[ 2 -7  2  2  2  2  2 -7  2] # ndfilters.convolve
[ 2 -7  2  2  2  2  2 -7  2] # expected result

1D chunked :
[ 2 -5  2  2  2  2  2 -4  2] # ndfilters.convolve
[ 2 -7  2  2  2  2  2 -7  2] # expected result

2D no chunks axis 0  :

[[ 10  10  10  10  10]
 [-15 -15 -15 -15 -15]
 [ 10  10  10  10  10]
 [-15 -15 -15 -15 -15]
 [ 10  10  10  10  10]] # ndfilters.convolve

[[ 10  10  10  10  10] 
 [-15 -15 -15 -15 -15]
 [ 10  10  10  10  10]
 [-15 -15 -15 -15 -15]
 [ 10  10  10  10  10]] # expected result


2D no chunks axis 1  :

[[ 2  2  2  2  2]
 [-3 -3 -3 -3 -3]
 [ 2  2  2  2  2]
 [-3 -3 -3 -3 -3]
 [ 2  2  2  2  2]] # ndfilters.convolve

[[ 2  2  2  2  2]
 [-3 -3 -3 -3 -3]
 [ 2  2  2  2  2]
 [-3 -3 -3 -3 -3]
 [ 2  2  2  2  2]] # expected result


2D chunked axis 0  :

[[ 10  10  10  10  10]
 [-15 -15 -15 -15 -15]
 [ 10  10  10  10  10]
 [-15 -15 -15 -15 -15]
 [ 10  10  10  10  10]] # ndfilters.convolve

[[ 10  10  10  10  10]
 [-15 -15 -15 -15 -15]
 [ 10  10  10  10  10]
 [-15 -15 -15 -15 -15]
 [ 10  10  10  10  10]] # expected result


2D chunked axis 1  :

[[ 2  2  2  2  2]
 [-2 -2 -2 -2 -2]
 [ 2  2  2  2  2]
 [-3 -3 -3 -3 -3]
 [ 2  2  2  2  2]] # ndfilters.convolve

[[ 2  2  2  2  2]
 [-3 -3 -3 -3 -3]
 [ 2  2  2  2  2]
 [-3 -3 -3 -3 -3]
 [ 2  2  2  2  2]] # expected result
```

**Anything else we need to know?**:

The workaround is used there to get the correct value might be a good way to fix this bug. 

**Environment**:

- Dask version: dask         2021.4.1; dask-core                 2021.4.1; dask-image                0.6.0              
- Python version: 3.9.5
- Operating System: Ubuntu 20
- Install method (conda, pip, source): dask and dask-core: pip; dask-image: conda-forge

**Comment by anlavandier** [comment]
After testing the same script with the other modes, namely `reflect`, `nearest` and `constant` (there's no equivalent to `mirror` in `dask.array.overlap.overlap`) everything works fine.

**Comment by GenevieveBuckley** [comment]
Thank you for the report!

**Comment by jakirkham** [comment]
What `convolve` is this being compared to? Is this using `scipy.ndimage` or something else?

**Comment by anlavandier** [comment]
> What `convolve` is this being compared to? Is this using `scipy.ndimage` or something else?

This is compared against itself but:
* For `O1` we only have `O1 = convolve(F,kernel, mode = 'wrap')`. The boundary condition (`mode = wrap` in our case) is left for `dask_image.ndfilters.convolve` to deal with.
* For `O2`, the boundary condition is dealt with beforehand by `dask.array.overlap.overlap` then we calculate the convolution using `dask_image.ndfilters.convolve` and finally remove the ghost zones that `dask.array.overlap.overlap` added using `dask.array.overlap.trim_overlap`. This refers to these lines in the script:
    * `F_1 = da.overlap.overlap(F, depth = {axis : 2}, boundary = 'periodic')`
    * `O2_int = convolve(F_1,kernel, mode = 'constant')`
    * `O2 = da.overlap.trim_overlap(O2_int, depth = {axis : 2})`

**Comment by jakirkham** [comment]
Sorry let me rephrase, where does the expected result come from? Is this based on the result of another library (like SciPy or scikit-image) or is this intuition?

**Comment by anlavandier** [comment]
The expected result is `O2`. In my examples it is very possible to calculate it by hand to verify that `O2` is the right value but I can update the script to add another instance of convolution to validate the results if you think that it is necessary.

**Comment by jakirkham** [comment]
Part of the reason I ask is under-the-hood we are using `scipy.ndimage.filters.convolve`. So a natural question is do we see the same behavior from SciPy alone?

**Comment by anlavandier** [comment]
> So a natural question is do we see the same behavior from SciPy alone?

I am not sure to understand what you are referring to in that case. The chunking problem cannot appear in SciPy because it doesn't support chunked arrays anyways. As far as correctness goes, I changed a bit of my script to show consistency with `scipy.ndimage.filters.convolve` instead of plain results. 

Script: 
```python
import numpy as np
import dask.array as da
from dask_image.ndfilters import convolve as di_cv
from scipy.ndimage import convolve as sndi_cv

def test_correct(F,axis = 0):
    if axis !=0:
        F = da.swapaxes(F, 0, axis)
        result = test_correct(F,axis = 0)
        return result
    
    kernel = np.array([1,-1,0,1,-1]).reshape((5,) + (1,) * (F.ndim - 1))
    # The output using ndfilters.convolve with the wrap mode
    O1 = di_cv(F,kernel, mode = 'wrap')
    # Output created by forcing the wrapping using dask then calculating the convolution. 
    # taking care of the boundary
    F_1 = da.overlap.overlap(F, depth = {axis : 2}, boundary = 'periodic')
    # calculate the convolution
    O2_int = di_cv(F_1,kernel, mode = 'constant')
    # removing the useless ghosts zones. 
    O2 = da.overlap.trim_overlap(O2_int, depth = {axis : 2})

    O3 = sndi_cv(F.compute(), kernel, mode  = "wrap")
    print("Plain dask_image.ndfilters.convolve vs Scipy :", np.allclose(O1.compute(), O3))
    print("dask.array.overlap  + dask_image.nfilters.convolve vs Scipy :", np.allclose(O2.compute(),O3), end = '\n\n')


test_1D_no_chunks = da.from_array([1,2,3,4,5,6,7,8,9],chunks = (-1,))
test_1D_chunked = da.from_array([1,2,3,4,5,6,7,8,9],chunks = (5,))

print('1D no chunks :')
test_correct(test_1D_no_chunks)
print('1D chunked :')
test_correct(test_1D_chunked)


test_2D_no_chunks = da.from_array(np.arange(25).reshape(5,5),chunks =(-1,-1))
test_2D_chunked = da.from_array(np.arange(25).reshape(5,5),chunks =(-1,2))

print('2D no chunks axis 0  :')
test_correct(test_2D_no_chunks,axis = 0)
print('2D no chunks axis 1  :')
test_correct(test_2D_no_chunks,axis = 1)

print('2D chunked axis 0  :')
test_correct(test_2D_chunked,axis = 0)
print('2D chunked axis 1  :')
test_correct(test_2D_chunked,axis = 1)
```
Results:
```python
 1D no chunks :
Plain dask_image.ndfilters.convolve vs Scipy : True
dask.array.overlap  + dask_image.nfilters.convolve vs Scipy : True

1D chunked :
Plain dask_image.ndfilters.convolve vs Scipy : False
dask.array.overlap  + dask_image.nfilters.convolve vs Scipy : True

2D no chunks axis 0  :
Plain dask_image.ndfilters.convolve vs Scipy : True
dask.array.overlap  + dask_image.nfilters.convolve vs Scipy : True

2D no chunks axis 1  :
Plain dask_image.ndfilters.convolve vs Scipy : True
dask.array.overlap  + dask_image.nfilters.convolve vs Scipy : True

2D chunked axis 0  :
Plain dask_image.ndfilters.convolve vs Scipy : True
dask.array.overlap  + dask_image.nfilters.convolve vs Scipy : True

2D chunked axis 1  :
Plain dask_image.ndfilters.convolve vs Scipy : False
dask.array.overlap  + dask_image.nfilters.convolve vs Scipy : True
```

**Comment by anlavandier** [comment]
The issue is now closed with the merging of #243
